[WIP] Fix dev environment memory leak

tenderlove opened the pull request:
This commit fixes the memory leak in dev.  I'm not super pleased with the commit right now and I want to refactor it some, but I also want to make sure it's public.  Specifically I dislike the part where I added another layer to the cache.  I think I can improve that part.

I think the source of the leak is that the digest cache was combined with the template object cache.  This commit splits the caches, then combines them behind the `DetailsKey` interface.  Resolvers have to ask the details key thing for their cache.  That gave the LookupContext an implicit dependency on `DetailsKey`, which is why I had to change the `fallbacks` method to do a check at runtime (otherwise the circular dependency ruins file loading).

I think separating the caches will fix the tests here: https://github.com/rails/rails/pull/35160

Here's a graph of the heap live slots on master versus this branch:

<img width="490" alt="erb eval 2019-02-07 15-48-58" src="https://user-images.githubusercontent.com/3124/52450461-eb9d0f80-2aef-11e9-9446-263ee5bfb539.png">

tenderlove committed:
Fix memory leaks in dev

This commit splits the digest cache from the template cache then makes
the template cache "per resolver".  This cache is stored in the
"details" object, so whenever we clear that, the cache will be cleared
(which is currently on every request).

rafaelfranca commented on actionview/lib/action_view/lookup_context.rb:
```diff
@@ -55,21 +61,50 @@ module Accessors #:nodoc:
+      # preallocate all the default blocks for performance/memory consumption reasons
+      PARTIAL_BLOCK = lambda { |cache, partial| cache[partial] = SmallCache.new }
+      PREFIX_BLOCK  = lambda { |cache, prefix|  cache[prefix]  = SmallCache.new(&PARTIAL_BLOCK) }
+      NAME_BLOCK    = lambda { |cache, name|    cache[name]    = SmallCache.new(&PREFIX_BLOCK) }
```
Whose are unaligned now, since we moved them I'd just remove the alignment

rafaelfranca commented on actionview/lib/action_view/lookup_context.rb:
```diff
@@ -55,21 +61,50 @@ module Accessors #:nodoc:
+          details = details.dup
+          details[:formats] &= Template::Types.symbols
+        end
+        @details_keys[details] ||= Object.new
```
It is not clear to my why the cache key is an instance of object. It is just to create a unique value?

tenderlove replied:
Yes.  I think the original reason was that calculating a hash for the details key was expensive (essentially `{ some: :complex, thing: { a: b: } }.hash`), so we do it once and use an object instead.  Apparently we have to consult the template cache many times, so we need a fast `.hash`.

I think we can eliminate this cache all together and still maintain speed, but it may require changes to the lookup context.  That's why I'm interested in making the lookup context more private (or be more aggressive with API changes).

rafaelfranca replied:
Makes sense.

rafaelfranca reviewed (commented):
Nice find. It looks goof to me so far

tenderlove committed:
Pass the template cache in to the finder

This means we only do one lookup with the details hash, and the cost of
that lookup can be amortized (rather than passing a special object
around)

tenderlove committed:
Remove references from the template cache

This other cache doesn't need to have references to the template cache
anymore because we pass the cache in.

tenderlove committed:
Use more descriptive variable names
